### DISCUSSION TITLE
Juniper & Bson

### DISCUSSION THREAD

**Discussion by NeotenicPrimate** [graphql-rust/juniper#1041] [Q&A] [answered]
I am having problems integrating MongoDB and Juniper because of Bson crate compatibility (MongoDB driver uses 2.0 while Juniper uses 1.0). I was told that 0.16 should solve this. I absolutely adore Juniper and would really like to use it in an upcoming project. Does anyone know when this will be released?

**Comment by Teajey** [comment]
I've been using `mongodb = "2.0"` and `juniper = "0.15"` together quite happily. What is it particularly that you're having trouble with integrating?

**Reply by NeotenicPrimate** [reply]
> I've been using `mongodb = "2.0"` and `juniper = "0.15"` together quite happily. What is it particularly that you're having trouble with integrating?

I am having problems converting str id into an OID that is accepted by the MongoDB driver, for example:

```
let str_id = "6230f0ecea0bc0db2d1e056a";
 let oid: ObjectId = bson::oid::ObjectId::with_string(str_id).unwrap();

  let movie = typed_collection
    .find_one(Some(doc! { "_id": oid }), None)
  // .find_one(Some(doc! { "title": "The Grapes of Wrath" }), None)
    .await
    .unwrap()
    .unwrap();

  println!("{:?}", movie);
```

returns:

```
error[E0277]: the trait bound `mongodb::bson::Bson: std::convert::From<bson::oid::ObjectId>` is not satisfied
  --> src/main.rs:82:20
   |
82 |     .find_one(Some(doc! { "_id": oid }), None)
   |                    ^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<bson::oid::ObjectId>` is not implemented for `mongodb::bson::Bson`
   |
```

**Reply by Teajey** [reply]
Do you need `bson::oid::ObjectId`? I just pass a string straight in
```rs
...
        id: Option<String>,
        limit: Option<i32>,
        skip: Option<i32>,
    ) -> FieldResult<Items<BlogPost>> {
        let collection = data::get_mongo_collection::<BlogPost>(&context.config, "blog_posts")?;

        let blog_posts;
        let mut length: Option<i32> = None;

        if let Some(id) = &id {
            let blog_post =
                collection
                    .find_one(doc! {"_id": id}, None)?
...
```

**Reply by NeotenicPrimate** [reply]
I do not. But when I try passing the string it returns a None (as in no result):
```
let str_id = "6230e8233c15e988a486c689".to_string();
// let oid: ObjectId = bson::oid::ObjectId::with_string(str_id).unwrap();

let movie = typed_collection
    .find_one(Some(doc! { "_id": str_id }), None)
    .await
    .unwrap()
    .unwrap();
```

**Reply by Teajey** [reply]
Are you having MongoDB generate the IDs for you? I just generate my own IDs so that I can use plain strings instead of something like `_id: ObjectId("623287aaa427e5f463766bc6")`, in which case I think `bson::oid::ObjectId` is required. But one option might be to see if there are settings for `Cargo.toml` where you can force MongoDB and Juniper to resolve to the same version of Bson. I know that this is possible with NPM, but I'm not as familiar with Cargo

**Comment by tyranron** [comment, marked as answer]
@NeotenicPrimate sorry, still no ETAs for `juniper 0.16`. We (active contributors) don't have enough bandwidth to give some guarantees about that, especially last three weeks having the war in our region.

**Reply by NeotenicPrimate** [reply]
> @NeotenicPrimate sorry, still no ETAs for `juniper 0.16`. We (active contributors) don't have enough bandwidth to give some guarantees about that, especially last three weeks having the war in our region.

Absolutely no problem, I was just curious. This is superficial and useless but I am wholeheartedly with you. Thank you for all you do.
